### ISSUE TITLE
dev_timeout seems to be ignored

### ISSUE THREAD

**Issue by AlpineScrewdriver** [Juniper/salt#124]
### Description of Issue/Question

SRX300/Virtual Chassis (slower commit scenarios) seem to fail most of the time, with a variety of errors that all seem related to timeouts. Only the junos.cli module seems to be relatively reliable at the moment. Unfortunately can't use salt in production with these issues.

### Setup
It is a fresh ubuntu 18.04 box with updates done
I installed salt this way:
sudo sh bootstrap.sh -Mg https://github.com/Juniper/salt.git
sudo pip install junos-eznc jxmlease

minion and master on the same box

/srv/pillar/srx300.sls
proxy:
  proxytype: junos
  host: 1.2.3.4
  username: user
  port: 830
  password: pw
  timeout: 300

/srv/pillar/top.sls
base:
  'srx300':
    - srx300

### Steps to Reproduce Issue
Attempt to commit/rollback a slower device (virtual chassis, srx300, etc).

### Versions Report
$ salt --versions-report
Salt Version:
           Salt: 2019.2.0
 
Dependency Versions:
           cffi: 1.12.3
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.3
      gitpython: 2.1.8
          ioflo: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.7
   msgpack-pure: Not Installed
 msgpack-python: 0.6.1
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.15rc1 (default, Nov 12 2018, 14:31:15)
   python-gnupg: 0.4.1
         PyYAML: 5.1
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5
 
System Versions:
           dist: Ubuntu 18.04 bionic
         locale: UTF-8
        machine: x86_64
        release: 4.18.0-18-generic
         system: Linux
        version: Ubuntu 18.04 bionic

Error output:

commit rpc was sent but device canceled
$ sudo salt 'srx300' junos.rollback id=1 dev_timeout=300
srx300:
    ----------
    message:
        Rollback successful but commit failed with error "RpcTimeoutError(host: 1.2.3.4, cmd: commit-configuration, timeout: 30)"
    out:
        False

commit rpc was sent but device canceled
$ sudo salt 'srx300' junos.rollback id=1 dev_timeout=300
srx300:
    ----------
    message:
        Rollback successful but commit failed with error "ConnectClosedError(1.2.3.4)"
    out:
        False

$ sudo salt 'srx300' junos.rollback id=1 dev_timeout=300
srx300:
    ----------
    message:
        Could not commit check due to "RpcTimeoutError(host: 1.2.3.4, cmd: commit-configuration, timeout: 30)"
    out:
        False

device did commit this time, but several seconds after task ended
$ date; sudo salt 'srx300' junos.rollback id=1 dev_timeout=300; date
Fri May 10 20:13:30 CDT 2019
srx300:
    ----------
    message:
        Rollback successful but commit failed with error "RpcTimeoutError(host: 1.2.3.4, cmd: commit-configuration, timeout: 30)"
    out:
        False
Fri May 10 20:14:30 CDT 2019

Attemping to push a larger config for testing, did not see a commit attempt on the device, only saw a rollback operation:

$ date; sudo salt 'srx300' junos.install_config 'salt://conf.set' dev_timeout=300; date
Fri May 10 20:24:00 CDT 2019
srx300:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1660, in _thread_return
        return_data = minion_instance.executors[fname](opts, data, func, args, kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/executors/direct_call.py", line 12, in execute
        return func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/junos.py", line 928, in install_config
        ret['out'] = False
      File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/utils/config.py", line 837, in __exit__
        self.unlock()
      File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/utils/config.py", line 597, in unlock
        raise err
    ConnectClosedError: ConnectClosedError(1.2.3.4)
ERROR: Minions returned with non-zero exit code
Fri May 10 20:24:45 CDT 2019

This time it did commit successfully, but still see an error:

$ date; sudo salt 'srx300' junos.install_config 'salt://conf.set' dev_timeout=300; date
Fri May 10 20:25:37 CDT 2019
srx300:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1660, in _thread_return
        return_data = minion_instance.executors[fname](opts, data, func, args, kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/executors/direct_call.py", line 12, in execute
        return func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/junos.py", line 928, in install_config
        ret['out'] = False
      File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/utils/config.py", line 837, in __exit__
        self.unlock()
      File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/utils/config.py", line 595, in unlock
        raise UnlockError(rsp=err.rsp)
    UnlockError: UnlockError
ERROR: Minions returned with non-zero exit code
Fri May 10 20:26:49 CDT 2019

**Comment by rsmekala** [comment]
@Zapotomous All the fixes have gone into Juniper/salt@2019.2, they are not present in the repo's master branch. Can you please try to check out salt from 2019.2 branch ??

Can you change the command and give it a try ??

Command: `sudo sh bootstrap.sh -Mg https://github.com/Juniper/salt.git@2019.2`
I don't know the exact command to install salt this way, it might be similar to the above command.

**Comment by AlpineScrewdriver** [comment]
Same problem. The unlock errors confuse me because I can run junos.lock and junos.unlock on the same device without issue. Works 100% of the time.

$ salt 'switchstack' junos.install_config 'salt://dns.conf' dev_timeout=300
switchstack:
    The minion function caused an exception: Traceback (most recent call last):
      File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 1660, in _thread_return
        return_data = minion_instance.executors[fname](opts, data, func, args, kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/executors/direct_call.py", line 12, in execute
        return func(*args, **kwargs)
      File "/usr/lib/python2.7/dist-packages/salt/modules/junos.py", line 928, in install_config
        ret['out'] = False
      File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/utils/config.py", line 837, in __exit__
        self.unlock()
      File "/usr/local/lib/python2.7/dist-packages/jnpr/junos/utils/config.py", line 595, in unlock
        raise UnlockError(rsp=err.rsp)
    UnlockError: UnlockError
ERROR: Minions returned with non-zero exit code

**Comment by AlpineScrewdriver** [comment]
Testing further, it does not seem to be slow device specific. When I run a junos.commit, sometimes I get a RPCTimeout Error 2 near immediately. If it passes that point, then I get a timeout on the commit check (30 seconds).

salt 'switchstack' junos.commit
switchstack:
    ----------
    message:
        Could not perform commit check due to "RpcTimeoutError(host: 1.2.3.4, cmd: commit-configuration, timeout: 2)"
    out:
        False

If I try to pass dev_timeout to junos.commit 

salt 'switchstack' junos.commit dev_timeout=300
switchstack:
    ----------
    message:
        Could not perform commit check due to "'Device' object has no attribute 'cu'"
    out:
        False

**Comment by AlpineScrewdriver** [comment]
So after looking into it further, I don't believe that command was actually installing 2019.2. 

From https://github.com/saltstack/salt-bootstrap#bootstrapping-salt:

To install a specific branch from a Git fork:

curl -o bootstrap-salt.sh -L https://bootstrap.saltstack.com
sudo sh bootstrap-salt.sh -g https://github.com/myuser/salt.git git mybranch

I used:
sudo sh bootstrap-salt.sh -Mg https://github.com/Juniper/salt.git git 2019.2

I am curious how others are installing salt if not via bootstrap. If there is a better way to go about this may want to consider adding to the readme. After installing salt this way I no longer see the timeouts so dev_timeout does seem to work. Closing.
